[FSDP] Post-bwd prefetch with reverse pre-forward order

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#84085 [FSDP] Post-bwd prefetch with reverse pre-forward order**
* #83917 [FSDP] Add rate limiter
* #84366 [FSDP][Easy] Save unpadded/padded unsharded sizes as attributes
* #83665 [FSDP] Generalize prefetching; lower unshard/reshard to handle
* #84761 [FSDP][Easy] Minor cleanup
* #84601 [FSDP] Subtest prefetching for `test_fsdp_grad_acc.py`
* #84600 [FSDP] Remove `forward_prefetch`
* #84598 [FSDP][Easy] Remove unused functions

This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

Commit:
[FSDP] Post-bwd prefetch with reverse pre-forward order

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

[ghstack-poisoned]

zhaojuanmao commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -483,12 +497,12 @@ def record_pre_forward(self, handles: List[FlatParamHandle]) -> None:
+            if self.is_first_iter:
+                self.handles_debug_pre_forward_order.append(handles_key)
+        # Only record the first usage of a handles key
+        if handles_key in self.handles_to_post_forward_order_index:
```
in 'handles_to_pre_forward_order_index'? why not record the order every iteration from the scratch? it is good for dynamic graph?

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Post-bwd prefetch with reverse pre-forward order"


This changes `BACKWARD_POST` prefetching to use the reverse pre-forward order. This needs benchmarking still.

The idea is that now we maintain a separate list `handles_debug_pre_forward_order` that is only modified on the 1st iteration and only used if the debug level is at least `INFO`. Then, `handles_pre_forward_order` and `handles_post_forward_order` are both modified per iteration in the forward to be used in the backward pass prefetching.

[ghstack-poisoned]

zhaojuanmao commented:
seems this PR is mixed with the 'use_orig_params' PR

awgu commented:
> seems this PR is mixed with the 'use_orig_params' PR

Ah. I messed up the rebase somehow. Let me pull these apart tomorrow.
